File: nvpl_blas/c/dtrmv.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_dtrmv
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 2;
    nvpl_int_t lda = 2;
    enum CBLAS_ORDER order = CblasRowMajor;
    enum CBLAS_TRANSPOSE transA = CblasNoTrans;
    enum CBLAS_UPLO uplo = CblasUpper;
    enum CBLAS_DIAG diag = CblasNonUnit;
    double * A = NULL;
    double * X;
    nvpl_int_t incX = 1;

    printf("\nExample: cblas_dtrmv for the triangular matrix-vector multiplication\n\n");
    printf("#### args: n=%" PRId64 ", lda=%" PRId64 ", incx=%" PRId64 ", transA=%c, order=%c, uplo=%c"
            ", diag=%c\n", (int64_t)N, (int64_t)lda, (int64_t)incX, transpose_to_char(transA), 
            order_to_char(order), uplo_to_char(uplo), diag_to_char(diag));

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);

    // allocate memory
    A = (double *)malloc(lda * N * sizeof(double));
    X = (double *)malloc(len_x * sizeof(double));

    // fill data
    fill_dmatrix(A, N, N, lda, order, UPLO_TO_FILL_MODE(uplo), diag);
    fill_dvector(X, N, incX);

    // print input data
    print_dmatrix(A, N, N, lda, "A", order);
    print_dvector(X, N, incX, "X");

    // call cblas_dtrmv
    cblas_dtrmv(order, uplo, transA, diag, N, A, lda, X, incX);

    // print result
    printf("\nAfter dtrmv operation:\n");
    print_dvector(X, N, incX, "X");

    // release memory
    free(A);
    free(X);
    return 0;
}

File: nvpl_blas/f77/dtrmv.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DTRMV 
!
!******************************************************************************
      program   DTRMV_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic  abs

      character*1      uplo, transa, diag
      parameter        (uplo='U', transa='N', diag='N')
      integer          n, lda
      integer          incx, i
      parameter        (n=2, lda=2, incx=1)
      double precision a(lda,n)
      integer          xsize
      parameter        (xsize=1+(n-1)*abs(incx))
      double precision x(xsize)
      data             (x(i), i=1,xsize)/1.0,-3.7/

*       External Subroutines
      external        print_dmatrix, fill_dmatrix, print_dvector
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, lda, incx, uplo, transa, diag
      call fill_dmatrix(a,n,n,lda,uplo,diag)
      call print_dmatrix(a,n,n,lda,'A')
      call print_dvector(x,n,incx,'X')
*
*      Call DTRMV subroutine
      call DTRMV(uplo,transa,diag,n,a,lda,x,incx)
*
      print*
      print 102, 'After dtrmv operation:'
      call print_dvector(x,n,incx,'X')

      stop
 99   format('Example: DTRMV for the triangular matrix-vector '
     &       'multiplication')
 100  format('#### args: n=',i1,', lda=',i1,', incx=',i1,', uplo=',a1,
     &       ', transa=',a1,', diag=',a1)
 102  format(A)
      end
